### PR TITLE
Upgrade mPDF to 8.1.4

### LINKED ISSUE #474: "Undefined array key" error message when exporting page with interwiki links
Hello,
When trying to generate a PDF using DW2PDF, I'm receiving the following error message:

````
Warning: Undefined array key 1 in /home/robertinho/public_html/wiki/lib/plugins/dw2pdf/vendor/mpdf/mpdf/src/Mpdf.php on line 8580

Warning: Trying to access array offset on value of type null in /home/robertinho/public_html/wiki/lib/plugins/dw2pdf/vendor/mpdf/mpdf/src/Mpdf.php on line 8580

Warning: Undefined array key 0 in /home/robertinho/public_html/wiki/lib/plugins/dw2pdf/vendor/mpdf/mpdf/src/Mpdf.php on line 6947

Warning: Trying to access array offset on value of type null in /home/robertinho/public_html/wiki/lib/plugins/dw2pdf/vendor/mpdf/mpdf/src/Mpdf.php on line 6947

Warning: Undefined array key 1 in /home/robertinho/public_html/wiki/lib/plugins/dw2pdf/vendor/mpdf/mpdf/src/Mpdf.php on line 8580

Warning: Trying to access array offset on value of type null in /home/robertinho/public_html/wiki/lib/plugins/dw2pdf/vendor/mpdf/mpdf/src/Mpdf.php on line 8580

Warning: Undefined array key 4 in /home/robertinho/public_html/wiki/lib/plugins/dw2pdf/vendor/mpdf/mpdf/src/Mpdf.php on line 8580

Warning: Trying to access array offset on value of type null in /home/robertinho/public_html/wiki/lib/plugins/dw2pdf/vendor/mpdf/mpdf/src/Mpdf.php on line 8580

Warning: Undefined array key 1 in /home/robertinho/public_html/wiki/lib/plugins/dw2pdf/vendor/mpdf/mpdf/src/Mpdf.php on line 6947

Warning: Trying to access array offset on value of type null in /home/robertinho/public_html/wiki/lib/plugins/dw2pdf/vendor/mpdf/mpdf/src/Mpdf.php on line 6947

Warning: Undefined array key 1 in /home/robertinho/public_html/wiki/lib/plugins/dw2pdf/vendor/mpdf/mpdf/src/Mpdf.php on line 6947

Warning: Trying to access array offset on value of type null in /home/robertinho/public_html/wiki/lib/plugins/dw2pdf/vendor/mpdf/mpdf/src/Mpdf.php on line 6947

Warning: Undefined array key 1 in /home/robertinho/public_html/wiki/lib/plugins/dw2pdf/vendor/mpdf/mpdf/src/Mpdf.php on line 6947

Warning: Trying to access array offset on value of type null in /home/robertinho/public_html/wiki/lib/plugins/dw2pdf/vendor/mpdf/mpdf/src/Mpdf.php on line 6947
````

The ``README.md`` file from DW2PDF states that mPDF is compatible up to PHP 7.4, but the mPDF itself already supports PHP 8.

I'm running DokuWiki Release 2022-07-31a "Igor" (stable) with PHP 8.1.

I'm attaching the markdown page (DokuWiki syntax) and the HTML page generated by the plugin as I was able to find the HTML into the ``data/cache`` folder. 

[start_dw2pdf_html.txt](https://github.com/splitbrain/dokuwiki-plugin-dw2pdf/files/10854790/start_dw2pdf_html.txt)
[start_markdown_dw.txt](https://github.com/splitbrain/dokuwiki-plugin-dw2pdf/files/10854791/start_markdown_dw.txt)

For me, this issue seems to be related with the number of images associated with the page. There is the interlink ``img`` tags and the number of errors seems to be twice as the number of images into the page (removing the ``img`` tag from the HTML file and trying to reparse it again decreases the number of errors).

### PR DESCRIPTION
**Pull request by eduardomozart**
Upgrade mPDF to 8.1.4 to add support for PHP 8.X as mPDF from DW2PDF supports PHP <= 7.4.
Sadly this breaks compatibility with PHP 7.4 because it's dependencies requires PHP >= 8.0.

### CODE REVIEW

**Review by Klap-in — CHANGES_REQUESTED** [review]
Thanks for the update.

Please check the manual patch, and reply it. It ensures the ACL are checked for the images from the wiki.

I observe that mpdf includes its own http client. A nice to have would be to configure it to use the http client of DokuWiki... see also https://mpdf.github.io/configuration/configuration-v7-x.html#http-client 

@splitbrain how do you judge dropping of php7.4 support? I guess it will result in questions..

File: vendor/mpdf/mpdf/src/Tag/Img.php [hunk 1]
```diff
@@ -3,7 +3,6 @@
 namespace Mpdf\Tag;
 
 use Mpdf\Mpdf;
-use \dokuwiki\plugin\dw2pdf\DokuImageProcessorDecorator as ImageProcessor;
```
[the remark this answers is not in the document]
Comment (eduardomozart) [reply vendor/mpdf/mpdf/src/Tag/Img.php]: I've applied this suggestion into this PR. Thank you!

File: vendor/mpdf/mpdf/src/Tag/Tag.php [hunk 2]
```diff
@@ -61,7 +61,7 @@ abstract class Tag
 	protected $colorConverter;
 
 	/**
-	 * @var ImageProcessor
+	 * @var \Mpdf\Image\ImageProcessor
```
[the remark this answers is not in the document]
Comment (eduardomozart) [reply vendor/mpdf/mpdf/src/Tag/Tag.php]: I've applied this suggestion into this PR. Thank you!

File: vendor/mpdf/mpdf/src/Tag/Tag.php [hunk 3]
```diff
@@ -8,7 +8,7 @@
 use Mpdf\Color\ColorConverter;
 use Mpdf\CssManager;
 use Mpdf\Form;
-use \dokuwiki\plugin\dw2pdf\DokuImageProcessorDecorator as ImageProcessor;
+use Mpdf\Image\ImageProcessor;
```
[the remark this answers is not in the document]
Comment (eduardomozart) [reply vendor/mpdf/mpdf/src/Tag/Tag.php]: I've applied this suggestion into this PR. Thank you!

**Comment by eduardomozart** [comment]
Thank you. I didn't know that, but I'd noticed that local images wasn't being loaded with a newer mPDF version and was trying to figure out why. Applied your suggestions and now DW2PDF seems to be working as expected.

**Comment by Klap-in** [comment]
Thanks for patching again. Could you check please why the unit tests did fail? Remarkably, it still throws an error about the decorator..

**Comment by eduardomozart** [comment]
Those unit tests seems to be failing because mPDF 8.1.4 requires PHP 8, which is the only test that succeeded as expected. It would be great to implement DokuWiki HTTPClient for mPDF, as it would support parameters set at the DokuWiki interface (like Proxy), but I don’t see developing it as a priority because I do not use any of those features right now (most of my bug contributions are related to bugs or features I found myself into our own Wiki environment). As developing isn’t my main function (I’m a SysAdmin), I can get in trouble in my work if I took to much time developing instead of deploying systems and attending our clients.

**Comment by Klap-in** [comment]
about http client, clear trade-off, thanks for the effort so far. If it would fit you, it would be an opportunity. Therefore, I brought it up. Without is also fine.

Ah, I see now something from composer is blocking it. That would mean for everybody with php7.4 and lower it is not usable. 
Since November 7.4 is also EOL. 

Feels a bit to early to completely not support php7 anymore, but we have also no method to indicate it to the user. Maybe the extension manager should get a way to inform users before installing.

**Pull request closed without merging**